File: Cpp/340.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        vector<int> m(128,0);
        int j=-1;
        int cnt=0;
        int res=0;
        for(int i=0;i<s.size();++i){
            cnt=cnt+(m[s[i]]==0);
            m[s[i]]++;
            
            while(cnt>k){
                j++;
                m[s[j]]--;
                cnt=cnt-(m[s[j]]==0);
            }
            res=max(res,i-j);
        }
        return res;
    }
};

int main(){

}

File: Python/340.py
class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if k == 0:
            return 0
        cnts = collections.defaultdict(int)
        maxLen = 0
        i = j = 0
        while j<len(s):
            cnts[s[j]] +=1
            j+=1
            while len(cnts) > k:
                cnts[s[i]] -=1
                if cnts[s[i]] == 0:
                    del cnts[s[i]]
                i+=1
            maxLen = max(maxLen, j-i)
        return maxLen
